File: src/22.括号生成.py
#
# @lc app=leetcode.cn id=22 lang=python3
#
# [22] 括号生成
# backtracking

from typing import List

# @lc code=start
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        g = self.generateParenthesis
        if n == 1:
            return ["()"]
        else:
            res = [ "(" + s + ")" for s in g(n-1)]
            for i in range(1, n):
                res += [x+y for x in g(i) for y in g(n-i)]
            return list(set(res))

# @lc code=end

s = Solution()
print(s.generateParenthesis(3))
print(s.generateParenthesis(4))

File: src/22.括号生成.go
package main

/*
 * @lc app=leetcode.cn id=22 lang=golang
 *
 * [22] 括号生成
 */
// backtracking

import "fmt"

// @lc code=start
func generateParenthesis(n int) []string {
	if n == 1 {
		return []string{"()"}
	}
	ans := make([]string, 0)
	for _, item := range generateParenthesis(n - 1) {
		ans = append(ans, "("+item+")")
	}
	for i := 1; i < n; i++ {
		for _, x := range generateParenthesis(i) {
			for _, y := range generateParenthesis(n - i) {
				ans = append(ans, x+y)
			}
		}
	}
	mp := make(map[string]int, 0)
	res := make([]string, 0)
	for _, item := range ans {
		l := len(mp)
		mp[item] = 0
		if len(mp) != l {
			res = append(res, item)
		}
	}
	return res
}

// @lc code=end

func main() {
	fmt.Println(generateParenthesis(3))
}
